src/OmniSharp.Host/MSBuild/Discovery/MSBuildLocator.cs

```diff
@@ -0,0 +1,150 @@
+                ImmutableArray.Create<MSBuildInstanceProvider>(
+                    new DevConsoleInstanceProvider(loggerFactory),
+                    new VisualStudioInstanceProvider(loggerFactory),
+                    new MonoInstanceProvider(loggerFactory),
+                    new StandAloneInstanceProvider(loggerFactory)));
+
+        public static MSBuildLocator CreateStandAlone(ILoggerFactory loggerFactory)
+            => new MSBuildLocator(loggerFactory,
+                ImmutableArray.Create<MSBuildInstanceProvider>(
+                    new StandAloneInstanceProvider(loggerFactory)));
+
+        public void RegisterInstance(MSBuildInstance instance)
+        {
+            if (_registeredInstance != null)
+            {
+                throw new InvalidOperationException("An MSBuild instance is already registered.");
+            }
+
+            _registeredInstance = instance ?? throw new ArgumentNullException(nameof(instance));
+
+            AppDomain.CurrentDomain.AssemblyResolve += Resolve;
+
+            if (instance.SetMSBuildExePathVariable)
+            {
+                var msbuildExePath = Path.Combine(instance.MSBuildPath, "MSBuild.exe");
+                var msbuildDllPath = Path.Combine(instance.MSBuildPath, "MSBuild.dll");
+
+                string msbuildPath = null;
+                if (File.Exists(msbuildExePath))
+                {
+                    msbuildPath = msbuildExePath;
+                }
+                else if (File.Exists(msbuildDllPath))
+                {
+                    msbuildPath = msbuildDllPath;
+                }
+
+                if (!string.IsNullOrEmpty(msbuildPath))
+                {
+                    Environment.SetEnvironmentVariable("MSBUILD_EXE_PATH", msbuildPath);
+                    _logger.LogInformation($"MSBUILD_EXE_PATH environment variable set to '{msbuildPath}'");
+                }
+                else
+                {
+                    _logger.LogError("Could not find MSBuild executable path.");
+                }
+            }
+
+            _logger.LogInformation($"Registered MSBuild instance: {instance}");
+        }
+
+        private Assembly Resolve(object sender, ResolveEventArgs e)
+        {
+            var assemblyName = new AssemblyName(e.Name);
+
+            _logger.LogDebug($"Attempting to resolve '{assemblyName}'");
+
+            if (s_msbuildAssemblies.Contains(assemblyName.Name))
+            {
+                var assemblyPath = Path.Combine(_registeredInstance.MSBuildPath, assemblyName.Name + ".dll");
+                var result = File.Exists(assemblyPath)
+                    ? Assembly.LoadFrom(assemblyPath)
+                    : null;
+
+                if (result != null)
+                {
+                    _logger.LogDebug($"Resolved '{assemblyName.Name}' to '{assemblyPath}'");
+                    return result;
+                }
+            }
+
+            return null;
```

Comment:
For my own edification, what happens if Resolve returns null?

Reply:
The runtime will continue to the next AssemblyResolve handler it finds. From https://msdn.microsoft.com/en-us/library/system.appdomain.assemblyresolve%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396:

> If more than one event handler is registered for this event, the event handlers are called in order until an event handler returns a value that isn't null. Subsequent event handlers are ignored.